simulation/halsim_ws_core/src/main/native/cpp/WSProvider_SimDevice.cpp

```diff
@@ -0,0 +1,148 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2017-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "WSProvider_SimDevice.h"
+
+#include <hal/Ports.h>
+
+namespace wpilibws {
+
+void HALSimWSProviderSimDevice::Initialize() {
+  HALSIM_RegisterSimValueCreatedCallback(
+      m_handle, this, HALSimWSProviderSimDevice::OnValueCreatedStatic, 1);
+}
+
+void HALSimWSProviderSimDevice::OnNetworkConnected(
```

Comment:
Also, the destructor should unregister the callbacks since it these providers can be freed.

Reply:
OK, just writing this out here so for my own clarity:

- Have `HALSimWSProviderSimDevices` also participate in the `OnNetworkConnected/OnNetworkDisconnected` events
- `HALSimWSProviderSimDevices` stores the websocket that it gets during the `OnNetworkConnected` call
- In the `DeviceCreatedCallback`, after we create the new `HALSimWSProviderSimDevice (no s)`, we'll call the individual `HALSimWSProviderSimDevice (no s)` s `OnNetworkConnected` with the websocket that the dummy provider has

- In `HALSimWSProviderSimDevice (no s)` s `OnNetworkConnected` event handler, if the incoming socket is different from the stored socket (e.g. started as null and now we're provided one either via `HALSimWSProviderSimDevices::DeviceCreatedCallback` or via `RegisterWebsocket` in the client/server implementations), we'll disconnect the existing one (if it isn't null), and then register the SimValueCreated callback + all the other stuff that's currently in `OnNetworkConnected`. If the incoming socket is the same, just bail (since we can assume that it should have been initialized by now)

Does that sound about right?

Reply:
Whoops also realized the JSON format that the SimDevice sends isn't consistent with the other providers, so that's something to fix too.

Reply:
Yes, that sounds right. Just remember to call `OnNetworkConnected` from the uv thread so that additional locking isn't required.